### PR TITLE
Support for KStream branch and split

### LINKED ISSUE #265: Support for KStream branch and split
Right now there is no way to bind multiple outbound KStream target types in KStream binder. We should provide support for branching and split into various output topics. 

See this: https://kafka.apache.org/0100/javadoc/org/apache/kafka/streams/kstream/KStream.html#branch(org.apache.kafka.streams.kstream.Predicate...)

See the original issue: https://github.com/spring-cloud/spring-cloud-stream/issues/1141

### PR DESCRIPTION
**Pull request by sobychacko**
Resolves #265

- Ensure that branching can be done and any message conversions
  done on all outbound data to multiple topics
- Refine MessageConversion logic in KStream binder
- Refactor message conversion code into a specific class
- Cleanup and refactoring
- Test to verify branching support
- More refinements in the way nativeEncoding/nativeDecoding logic works in KStream binder
- Doc changes for KStream binder

### CODE REVIEW

**Comment by sobychacko** [comment]
PR is updated with a new commit. This PR now requires the merging of https://github.com/spring-cloud/spring-cloud-stream/pull/1179 in core.

**Comment by xmlking** [comment]
Conforming that, I am able to test this PR and works for my needs.
Build this PR locally and tested with this sample app: https://github.com/xmlking/klogs/tree/master/classifier

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 1]
```diff
@@ -553,36 +554,145 @@ public class WordCountProcessorApplication {
...
+
+If the following property is set (default is false), the framework skips all message conversions on the outbound (producer) side and it is then done by Kafka itself.
+
+`spring.cloud.stream.bindings.output.producer.useNativeEncoding`.
+
+Similarly, if the following property is set (default is false), any message conversion is skipped on the inbound (consumer) side and natively done by Kafka.
+
+`spring.cloud.stream.bindings.input.consumer.useNativeDecoding`.
+
+When native encoding is disabled, then the messages on the outbound are converted by Spring Cloud Stream using the provided contentType.
```
Comment (olegz) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: I would reword it as "...provided `contentType` header"?

Comment (olegz) [reply spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: And basically every time you mention header name, property name, or value such as true/false, i would put them in some type of quotes. Although it's up to you

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 2]
```diff
@@ -553,36 +554,145 @@ public class WordCountProcessorApplication {
...
+  destination: words
+  consumer:
+    headerMode: raw
+----
 
 === Support for interactive queries
 
 If access to the `KafkaStreams` is needed for interactive queries, the internal `KafkaStreams` instance can be accessed via `KStreamBuilderFactoryBean.getKafkaStreams()`.
-You can autowire the `KStreamBuilderFactoryBean` instance provided by the KStream binder. Then you can get `KafkaStreams` instance from it and retrieve the underlying store, execute queries on it, etc.
+You can autowire the `KStreamBuilderFactoryBean` instance provided by the KStream binder.
```
Comment (olegz) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: I believe you or Gary mentioned that KStream is being renamed to 'Kafka Streams'. Should we use it here and put KStream in parentheses?

File: spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java [hunk 3]
```diff
@@ -80,42 +85,77 @@ public KStreamBinder(KafkaBinderConfigurationProperties binderConfigurationPrope
...
+			if (StringUtils.hasText(properties.getExtension().getKeySerde())) {
+				keySerde = Utils.newInstance(properties.getExtension().getKeySerde(), Serde.class);
+				if (keySerde instanceof Configurable) {
+					((Configurable) keySerde).configure(streamsConfig.originals());
 				}
+			}
+			else {
+				keySerde = this.binderConfigurationProperties.getConfiguration().containsKey("key.serde") ?
+						Utils.newInstance(this.binderConfigurationProperties.getConfiguration().get("key.serde"), Serde.class) : Serdes.ByteArray();
+			}
```
Comment (olegz) [inline spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java]: Can you raise an issue to revisit this? I am not fully grasping context here yet, but I just made similar change to our partitioning. Basically I would love to avoid `newInstance` in favor of making things Spring configured. It was easy for partitioning and should be easy here as well, but we can talk about it as separate improvement.

File: spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamListenerSetupMethodOrchestrator.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * Kafka Streams specific implementation for {@link StreamListenerSetupMethodOrchestrator}
+ * that overrides the default mechanisms for invoking StreamListener adapters.
+ *
+ * @author Soby Chacko
+ */
+public class KStreamListenerSetupMethodOrchestrator implements StreamListenerSetupMethodOrchestrator, ApplicationContextAware {
```
Comment (olegz) [inline spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamListenerSetupMethodOrchestrator.java]: Does it have to be public?

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc (hunk 1 above)
Comment (sobychacko) [reply spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: will do.

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc (hunk 2 above)
Comment (sobychacko) [reply spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: @olegz We need to go through another round of changes (in a separate PR) for renaming all these components. Better to rename everything in one go to avoid temporary inconsistencies.

File: spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java (hunk 3 above)
Comment (sobychacko) [reply spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java]: Issue created: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/288

File: spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamListenerSetupMethodOrchestrator.java (hunk 4 above)
Comment (sobychacko) [reply spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamListenerSetupMethodOrchestrator.java]: Because configuration beans reside in a different package, we need to make this implementation public.
